zerver/management/commands/deliver_scheduled_emails.py

```diff
@@ -34,23 +34,18 @@ class Command(BaseCommand):
 
     def handle(self, *args: Any, **options: Any) -> None:
         while True:
-            found_rows = False
             with transaction.atomic():
-                email_jobs_to_deliver = (
+                job = (
                     ScheduledEmail.objects.filter(scheduled_timestamp__lte=timezone_now())
                     .prefetch_related("users")
-                    .select_for_update()
+                    .select_for_update(skip_locked=True)
+                    .order_by("scheduled_timestamp")
+                    .first()
                 )
-                if email_jobs_to_deliver:
-                    found_rows = True
-                    for job in email_jobs_to_deliver:
-                        try:
-                            deliver_scheduled_emails(job)
-                        except EmailNotDeliveredException:
-                            logger.warning("%r not delivered", job)
-            # Less load on the db during times of activity,
-            # and more responsiveness when the load is low
-            if found_rows:
-                time.sleep(10)
-            else:
-                time.sleep(2)
+                if job:
+                    try:
+                        deliver_scheduled_emails(job)
+                    except EmailNotDeliveredException:
+                        logger.warning("%r not delivered", job)
+                else:
+                    time.sleep(2)
```

Comment:
The sleep logic here is a bit puzzling; I would expect us to want to sleep like 10s if there were no jobs, to avoid doing a database query every 2s on an idle Zulip server (which seems like a big resource waste).  I don't think our scheduled emails offer finer than minute-level granularity anyway.  It seems like the previous 2/10 logic was buggy/backwards, so no objections to removing it BTW.

But maybe we should instead just borrow the `ScheduledMessage` logic that just sleeps until the next minute?  Not sure:
```
    def handle(self, *args: Any, **options: Any) -> None:                                                
        try:                                                                                             
            while True:                                                                                  
                with transaction.atomic():                                                               
                    messages_to_deliver = ScheduledMessage.objects.filter(                               
                        scheduled_timestamp__lte=timezone_now(), delivered=False                         
                    ).select_for_update()                                                                
                    for message in messages_to_deliver:                                                  
                        do_send_messages([self.construct_message(message)])                              
                        message.delivered = True                                                         
                        message.save(update_fields=["delivered"])                                        
                                                                                                         
                cur_time = timezone_now()                                                                
                time_next_min = (cur_time + timedelta(minutes=1)).replace(second=0, microsecond=0)       
                sleep_time = (time_next_min - cur_time).total_seconds()                                  
                time.sleep(sleep_time)                                                                   
        except KeyboardInterrupt:                                                                        
            pass                                                                                         
```
@andersk thoughts?

Reply:
This seems like a good opportunity to think about whether we want to pull in [django-pg-queue](https://pypi.org/project/django-pg-queue/). It supports delayed jobs with `execute_at`, and I expect its use of `NOTIFY` will be both more accurate and more efficient than our polling loop.

We could also use `NOTIFY` ourselves, of course.

Reply:
I don't think load from this will ever be significant, so spikiness probably doesn't matter.

I am now thinking converting these two workers `django-pg-queue` would be a good way to get experience working with that library, since these are already in postgres.